scripting/srccoop_addon_effects_fix.sp

```diff
@@ -0,0 +1,192 @@
+#include <clientprefs>
+#include <sdkhooks>
+#include <sourcemod>
+#include <srccoop_api>
+
+#pragma newdecls required
+#pragma semicolon 1
+#pragma tabsize 0
+
+public Plugin myinfo =
+{
+	name        = "SourceCoop damage effects fix",
+	author      = "Jimbab",
+	description = "Client side damage effects fix for all weapons.",
+	version     = "1.1.0",
+	url         = "https://github.com/ampreeT/SourceCoop"
+};
+
+public void OnPluginStart()
+{
+	InitSourceCoopAddon();
+}
+
+public void OnEntityCreated(int entityIndex, const char[] szClassname)
+{
+	if (CBaseEntity(entityIndex).IsClassNPC())
+		SDKHook(entityIndex, SDKHook_OnTakeDamage, DispatchEffects);
+}
+
+public Action DispatchEffects(int victim, int& attacker, int& inflictor, float& damage, int& damagetype, int& weapon, float damageForce[3], float damagePosition[3])
+{
+	// Construct our entity objects based on their entity ID's
+	CBaseEntity playerAttacker = CBaseEntity(attacker);
+	CBaseEntity npcVictim = CBaseEntity(victim);
+
+	// Check victim and attacker are both npc and real player, respectively
+	if (!playerAttacker.IsClassPlayer() || !npcVictim.IsClassNPC())
+		return Plugin_Continue;
+
+	// Recreate the effects
+	CreateDamageEffect(damagePosition, npcVictim);
+
+	return Plugin_Continue;
+}
+
+/*
+ * Do our prop checks to check what type of entity we hit.
+ */
+void CreateDamageEffect(float position[3], CBaseEntity parent)
+{
+	int bloodColour = GetEntProp(parent.GetEntIndex(), Prop_Data, "m_bloodColor");
+
+	switch (bloodColour)
+	{
+		case (-1):    // Turrets, Mechanical, etc.
+		{
+			HandleEffects_Metal(position, parent);
+		}
+
+		case (0):    // Human NPC's
+		{
+			HandleEffects_RedBlood(position, parent);
+		}
+
+		case (1):    // Alien NPC's (yellow blood)
+		{
+			HandleEffects_YellowBlood(position, parent);
+		}
+
+		case (2):    // Alien NPC's (green blood)
+		{
+			HandleEffects_GreenBlood(position, parent);
+		}
+
+		case (3):    // NPC's with synthetic blood
+		{
+			HandleEffects_SynthBlood(position, parent);
+		}
+	}
+}
+
+/*
+ * Particle generator
+ */
+void CreateParticleEffect(float position[3], CBaseEntity parent, char[] effectName, float lengthTime = 5.0)
+{
+	CBaseEntity particle = CBaseEntity(CreateEntityByName("info_particle_system"));
+
+	if (particle.IsValid())
+	{
+		particle.Teleport(position);
+
+		particle.SetKeyValueStr("targetname", "blood_particle");
+		particle.SetKeyValueStr("effect_name", effectName);
+		particle.SetParent(parent);
+
+		particle.Spawn();
+		particle.Activate();
+
+		particle.AcceptInputStr("start");
+		CreateTimer(lengthTime, Callback_DeleteParticleSystem, particle);
+	}
+}
+
+/*
+ * Particle system deletion callback
+ */
+public Action Callback_DeleteParticleSystem(Handle timer, CBaseEntity particle)
+{
+	if (particle.IsValid())
+	{
+		char className[64];
+		particle.GetClassname(className, sizeof(className));
+
+		if (StrEqual(className, "info_particle_system", false))
```

Comment:
Why the class check?

Reply:
A better way to do entity removal without any potential incorrect class or index being passed would be to add it as an output and fire the output:

	char szRemoveOutput[64];
	Format(szRemoveOutput, sizeof(szRemoveOutput), "!self,kill,,%i,-1", lengthTime);

	particle.SetKeyValueStr("OnUser1", szRemoveOutput);
	particle.AcceptInput("FireUser1");

This will let the engine handle the entity removal and scheduling safely without the worry for it being removed before the timer runs out, or map change, or whichever that could change what index is passed.

Reply:
The main difference is allowing you to specify a time for it to fire and the engine will handle when that timer expires using the event queue instead of creating timers in SM. It will send the same input, it just does it through different means and hands it off to the game code.
Either method *should* work. The wrapper made in SourceCoop does alleviate a majority of cases by passing the CBaseEntity methodmap to the timer. But if all you need to do is fire a particular input to an entity after some time has passed, the safest way may be by OnUser# and FireUser#. If you need to modify a property or something like that, passing CBaseEntity would be one of the better ways to do that instead.